Generalize bilinear filler to for N-D multilinear filler

christianpayer opened the pull request:
This branch changes the bilinear filler (#2213) to support an arbitrary number of dimensions. Also, different scaling factors for each dimension are supported. I changed the name of the filler from "bilinear" to "linear", as it suits better for its multi-dimensional functionality.

The usage of this filler is the same as before (see comments, or #2213). Tests are not included, but I plan to write some using python scikit-image.

shelhamer commented:
While there is no reason this filler should be limited to 4-D, as it is now, this breaks backward compatibility as-is since there are existing definitions that make use of `bilinear`. See our [upgrade proto](https://github.com/BVLC/caffe/blob/master/src/caffe/util/upgrade_proto.cpp) code for an idea of how to automatically upgrade old definitions -- this is our practice for schema changes.

See #2213 for some earlier discussion of the naming. `linear` doesn't feel quite right since that could be nearly anything, although `bilinear` is admittedly over specific. Perhaps `interp` would do? The combination of a linear layer like `Convolution` and an `interp` filler might be clear enough.

Thoughts @longjon?

longjon commented:
In the usual interpolation sense (i.e., as a function of _coordinates_), _linear_ is not correct; _bilinear_ functions are _nonlinear_ functions which are separately linear in each of their arguments. `interp` is too generic, as there are non-bilinear interpolation methods (e.g., biquadratic or bicubic) which can be implemented using fillers and convolutions.

The generalization of bilinear is _multilinear_. Perhaps these could be accepted as synonyms.

christianpayer commented:
Thanks for your suggestions. I agree that `linear` is not the best name and that `multilinear` is much better.
If you agree to use `multilinear` I will change it and adapt the code for upgrade proto.

seanbell commented on include/caffe/filler.hpp:
```diff
@@ -229,32 +229,39 @@ layer {
   virtual void Fill(Blob<Dtype>* blob) {
-    CHECK_EQ(blob->num_axes(), 4) << "Blob must be 4 dim.";
-    CHECK_EQ(blob->width(), blob->height()) << "Filter must be square";
+    CHECK_GT(blob->num_axes(), 3) << "Blob must have at least 3 dimensions.";
```
Shouldn't this be `CHECK_GE`, if you want to encode "at least 3"?

christianpayer replied:
You are right, thanks for pointing that out. I will change it. I never tested the code for 1d inputs, but it should work, too.

seanbell commented on include/caffe/filler.hpp:
```diff
@@ -229,32 +229,39 @@ layer {
+        unsigned int sampling_factor = std::ceil(shape / 2.0f);
+        float center = (2 * sampling_factor - 1 - sampling_factor % 2)
+            / (2.0f * sampling_factor);
+        float coordinate = (i / stride) % shape;
```
Since this gets saved into `data` which has type `Dtype`, maybe the `float` should all be `Dtype`?

christianpayer replied:
Thanks, I agree. Using Dtype would remove the implicit conversion for weight in the following line.

seanbell commented on include/caffe/filler.hpp:
```diff
@@ -229,32 +229,39 @@ layer {
+      Dtype weight = 1;
+      for (int axis = 0; axis < blob->num_axes() - 2; ++axis) {
+        unsigned int shape = blob->shape(axis + 2);
+        unsigned int sampling_factor = std::ceil(shape / 2.0f);
```
Calling this `factor` instead of `sampling_factor` would better fit with the above block comment that explains how it works.

shelhamer commented:
@christianpayer I second `multilinear`, so if you agree then go ahead with the new name.
Once done, please squash your commits into a new commit with a message that reflects the new filler (in particular the new name and generality of dimensions). Thanks!

christianpayer commented:
@shelhamer Thanks for the comments. I changed the name to `multilinear`, squashed the commit and changed the commit message.
I did not edit upgrade_proto.cpp, as I am not sure, where/how to change the code there. As far as I can see an upgrade of a filler name has not been done before. Do you want me to implement something like
```
bool NetNeedsWeightFillerUpgrade(const NetParameter& net_param);
void UpgradeNetWeightFiller(NetParameter* net_param);
```
in upgrade_proto.hpp/cpp?

shelhamer commented:
@christianpayer adding the check `NetNeedsWeightFIllerUpgrade()` and update `UpgradeNetWeightFiller()` is the idea. You'll need to be a bit familiar with the C++ interface to protobuf, but hopefully the existing methods can be a guide. Here's a quick sketch to illustrate key pieces:

```
for (int i = 0; i < net_param->layer_size(); ++i) {
      if (net_param->layer(i).has_convolution_param()) {
          FillerParameter* filler_param = net_param->mutable_layer(i)->mutable_convolution_param()->mutable_filler_parameter();
          if (filler_param.type() == "bilinear") {
              filler_param->set_type("multilinear");
          }
      }
}
```

Care to take a stab at it and we'll go from there?

christianpayer commented:
@shelhamer Thanks for the code snippet. I just implemented upgrade_proto for weight fillers and hope that this is what you had in mind.

christianpayer commented:
Please do not merge this PR, as I implemented in the meantime a more general interpolation filler, that allows linear/cubic/lanczos interpolation and rewrites most of the code of this PR.
Next week I will create a new PR with this more general filler.

christianpayer commented:
Closed in favour of #4198.
